packages/contracts-bedrock/scripts/L2Genesis.s.sol

```diff
@@ -506,30 +522,11 @@ contract L2Genesis is Deployer {
     }
 
     /// @notice Sets all the preinstalls.
-    ///         Warning: the creator-accounts of the preinstall contracts have 0 nonce values.
-    ///         When performing a regular user-initiated contract-creation of a preinstall,
-    ///         the creation will fail (but nonce will be bumped and not blocked).
-    ///         The preinstalls themselves are all inserted with a nonce of 1, reflecting regular user execution.
-    function setPreinstalls() internal {
-        _setPreinstallCode(Preinstalls.MultiCall3);
-        _setPreinstallCode(Preinstalls.Create2Deployer);
-        _setPreinstallCode(Preinstalls.Safe_v130);
-        _setPreinstallCode(Preinstalls.SafeL2_v130);
-        _setPreinstallCode(Preinstalls.MultiSendCallOnly_v130);
-        _setPreinstallCode(Preinstalls.SafeSingletonFactory);
-        _setPreinstallCode(Preinstalls.DeterministicDeploymentProxy);
-        _setPreinstallCode(Preinstalls.MultiSend_v130);
-        _setPreinstallCode(Preinstalls.Permit2);
-        _setPreinstallCode(Preinstalls.SenderCreator_v060); // ERC 4337 v0.6.0
-        _setPreinstallCode(Preinstalls.EntryPoint_v060); // ERC 4337 v0.6.0
-        _setPreinstallCode(Preinstalls.SenderCreator_v070); // ERC 4337 v0.7.0
-        _setPreinstallCode(Preinstalls.EntryPoint_v070); // ERC 4337 v0.7.0
-        _setPreinstallCode(Preinstalls.BeaconBlockRoots);
-        _setPreinstallCode(Preinstalls.CreateX);
-        // 4788 sender nonce must be incremented, since it's part of later upgrade-transactions.
-        // For the upgrade-tx to not create a contract that conflicts with an already-existing copy,
-        // the nonce must be bumped.
-        vm.setNonce(Preinstalls.BeaconBlockRootsSender, 1);
+    function setPreinstalls() public {
+        address tmpSetPreinstalls = address(uint160(uint256(keccak256("SetPreinstalls"))));
+        vm.etch(tmpSetPreinstalls, vm.getDeployedCode("SetPreinstalls.s.sol:SetPreinstalls"));
+        SetPreinstalls(tmpSetPreinstalls).setPreinstalls();
+        vm.etch(tmpSetPreinstalls, "");
```

Comment:
I see why you use this pattern, but curious why you went with this over alternatives like inheriting from `SetPreinstalls`, or making `SetPreinstalls` a library that we call here?

Reply:
True, but we probably should start by optimizing for readable/maintainable and switch to workarounds like this once necessary—the `SetPreinstalls` contract is pretty simple, I would be surprised if duplicating its compilation makes a significant impact on compile times. Perhaps worth benchmarking?

Reply:
(also this is not a blocking comment)

Reply:
I'm fine going with another approach, I am not aware of any script composition dos/don'ts. Opened an issue to track this, and if we want to prioritize a particular style: https://github.com/ethereum-optimism/optimism/issues/11892

Reply:
Thanks, resolving this thread